tower-balance/src/lib.rs

```diff
@@ -283,3 +299,137 @@ impl<F: Future, E> Future for ResponseFuture<F, E> {
         self.0.poll().map_err(Error::Inner)
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use futures::future;
+    use quickcheck::*;
+    use tower_discover::Change;
+    use std::collections::VecDeque;
+
+    use super::*;
+
+    struct ReluctantDisco(VecDeque<Change<usize, ReluctantService>>);
+
+    struct ReluctantService {
+        polls_until_ready: usize,
+    }
+
+    impl Discover for ReluctantDisco {
+        type Key = usize;
+        type Request = ();
+        type Response = ();
+        type Error = ();
+        type Service = ReluctantService;
+        type DiscoverError = ();
+
+        fn poll(&mut self) -> Poll<Change<Self::Key, Self::Service>, Self::DiscoverError> {
+            let r = self.0.pop_front().map(Async::Ready).unwrap_or(Async::NotReady);
+            debug!("polling disco: {:?}", r.is_ready());
+            Ok(r)
+        }
+    }
+
+    impl Service for ReluctantService {
+        type Request = ();
+        type Response = ();
+        type Error = ();
+        type Future = future::FutureResult<(), ()>;
+
+        fn poll_ready(&mut self) -> Poll<(), ()> {
+            if self.polls_until_ready == 0 {
+                return Ok(Async::Ready(()));
+            }
+
+            self.polls_until_ready -= 1;
+            return Ok(Async::NotReady);
+        }
+
+        fn call(&mut self, _: ()) -> Self::Future {
+            future::ok(())
+        }
+    }
+
+    quickcheck! {
+        /// Creates a random number of services, each of which must be polled a random
+        /// number of times before becoming ready. As the balancer is polled, ensure that
+        /// it does not become ready prematurely and that services are promoted from
+        /// not_ready to ready.
+        fn poll_ready(service_tries: Vec<usize>) -> TestResult {
+            // Stores the number of pending services after each poll_ready call.
+            let mut pending_at = Vec::new();
+
+            let disco = {
+                let mut changes = VecDeque::new();
+                for (i, n) in service_tries.iter().map(|n| *n).enumerate() {
+                    for j in 0..n {
+                        if j == pending_at.len() {
+                            pending_at.push(1);
+                        } else {
+                            pending_at[j] += 1;
+                        }
+                    }
+
+                    let s = ReluctantService { polls_until_ready: n };
+                    changes.push_back(Change::Insert(i, s));
+                }
+                ReluctantDisco(changes)
+            };
+            pending_at.push(0);
+
+            let mut balancer = Balance::new(disco, choose::RoundRobin::default());
+
+            let services = service_tries.len();
+            let mut next_pos = 0;
+            for pending in pending_at.iter().map(|p| *p) {
+                assert!(pending <= services);
```

Comment:
Is there any particular reason for using an assertion here and `if ... { return TestResult::failed() }` elsewhere in the test?

Reply:
because this is expressing an invariant on the test logic, not validating app behavior